### PR TITLE
[DPAS] Lower the tt.dot to Intel DPAS instruction

### PR DESCRIPTION
**Pull request by chengjunlu**
Lower the tt.dot to Intel DPAS instruction.

- Modify the DpasEncodingAttr.
- Add a pass to accelerate matmul for Intel GPU.
- Add the TritonIntelGPU dialect for Intel ops/attr/passes on ttgir.

### CODE REVIEW

**Comment by chengjunlu** [comment]
The flash attention test case fails on DPAS.

```
2024-02-04T06:33:00.6611773Z =========================== short test summary info ============================
2024-02-04T06:33:00.6612665Z FAILED operators/test_flash_attention.py::test_op[True-True-dtype0-2-4-512-16] - AssertionError: Tensor-likes are not close!
2024-02-04T06:33:00.6613239Z 
2024-02-04T06:33:00.6613373Z Mismatched elements: 1528 / 65536 (2.3%)
2024-02-04T06:33:00.6613864Z Greatest absolute difference: 0.1552734375 at index (40966,) (up to 0.01 allowed)
2024-02-04T06:33:00.6614463Z Greatest relative difference: inf at index (711,) (up to 0 allowed)
2024-02-04T06:33:00.6615337Z FAILED operators/test_flash_attention.py::test_op[True-True-dtype1-2-4-512-16] - AssertionError: Tensor-likes are not close!
2024-02-04T06:33:00.6615896Z 
2024-02-04T06:33:00.6616015Z Mismatched elements: 13 / 65536 (0.0%)
2024-02-04T06:33:00.6616488Z Greatest absolute difference: 0.15625 at index (40966,) (up to 0.1 allowed)
2024-02-04T06:33:00.6617087Z Greatest relative difference: 2.65625 at index (40966,) (up to 0 allowed)
2024-02-04T06:33:00.6617985Z FAILED operators/test_flash_attention.py::test_op[True-False-dtype0-2-4-512-16] - AssertionError: Tensor-likes are not close!
2024-02-04T06:33:00.6618553Z 
2024-02-04T06:33:00.6618678Z Mismatched elements: 2873 / 65536 (4.4%)
2024-02-04T06:33:00.6619254Z Greatest absolute difference: 0.05438232421875 at index (59574,) (up to 0.01 allowed)
2024-02-04T06:33:00.6619868Z Greatest relative difference: inf at index (773,) (up to 0 allowed)
2024-02-04T06:33:00.6620733Z FAILED operators/test_flash_attention.py::test_op[True-False-dtype1-2-4-512-16] - AssertionError: Tensor-likes are not close!
2024-02-04T06:33:00.6621298Z 
2024-02-04T06:33:00.6621426Z Mismatched elements: 65536 / 65536 (100.0%)
2024-02-04T06:33:00.6621885Z Greatest absolute difference: nan at index (0,) (up to 0.1 allowed)
2024-02-04T06:33:00.6622431Z Greatest relative difference: nan at index (0,) (up to 0 allowed)
2024-02-04T06:33:00.6623278Z FAILED operators/test_flash_attention.py::test_op[False-True-dtype0-2-4-512-16] - AssertionError: Tensor-likes are not close!
2024-02-04T06:33:00.6623841Z 
2024-02-04T06:33:00.6624027Z Mismatched elements: 1526 / 65536 (2.3%)
2024-02-04T06:33:00.6624533Z Greatest absolute difference: 0.1553955078125 at index (40966,) (up to 0.01 allowed)
2024-02-04T06:33:00.6625137Z Greatest relative difference: inf at index (711,) (up to 0 allowed)
2024-02-04T06:33:00.6626011Z FAILED operators/test_flash_attention.py::test_op[False-True-dtype1-2-4-512-16] - AssertionError: Tensor-likes are not close!
2024-02-04T06:33:00.6626576Z 
2024-02-04T06:33:00.6626696Z Mismatched elements: 13 / 65536 (0.0%)
2024-02-04T06:33:00.6627160Z Greatest absolute difference: 0.15625 at index (40966,) (up to 0.1 allowed)
2024-02-04T06:33:00.6627744Z Greatest relative difference: 2.640625 at index (40966,) (up to 0 allowed)
2024-02-04T06:33:00.6628647Z FAILED operators/test_flash_attention.py::test_op[False-False-dtype0-2-4-512-16] - AssertionError: Tensor-likes are not close!
2024-02-04T06:33:00.6629209Z 
2024-02-04T06:33:00.6629340Z Mismatched elements: 2896 / 65536 (4.4%)
2024-02-04T06:33:00.6629833Z Greatest absolute difference: 0.05462646484375 at index (59574,) (up to 0.01 allowed)
2024-02-04T06:33:00.6630453Z Greatest relative difference: inf at index (773,) (up to 0 allowed)
2024-02-04T06:33:00.6631331Z FAILED operators/test_flash_attention.py::test_op[False-False-dtype1-2-4-512-16] - AssertionError: Tensor-likes are not close!
2024-02-04T06:33:00.6631890Z 
2024-02-04T06:33:00.6632113Z Mismatched elements: 65536 / 65536 (100.0%)
2024-02-04T06:33:00.6632611Z Greatest absolute difference: nan at index (0,) (up to 0.1 allowed)
2024-02-04T06:33:00.6633158Z Greatest relative difference: nan at index (0,) (up to 0 allowed)
2024-02-04T06:33:00.6633675Z ======================== 8 failed, 24 skipped in 50.80s ========================
```

**Comment by chengjunlu** [comment]
These matmul test cases only failed on PVC not on ATS.

```
=========================== short test summary info ============================
FAILED operators/test_matmul.py::test_op[128-128-32-1-4-2-107-233-128-False-False-bfloat16-bfloat16-True-True-None-None] - AssertionError: Tensor-likes are not close!

Mismatched elements: 384 / 24931 (1.5%)
Greatest absolute difference: 1.2578125 at index (61, 180) (up to 1e-05 allowed)
Greatest relative difference: 4.375 at index (47, 190) (up to 0.016 allowed)
FAILED operators/test_matmul.py::test_op[128-128-32-1-4-2-107-233-128-False-False-float16-float16-True-True-None-None] - AssertionError: Tensor-likes are not close!

Mismatched elements: 384 / 24931 (1.5%)
Greatest absolute difference: 1.2578125 at index (61, 180) (up to 1e-05 allowed)
Greatest relative difference: 4.3671875 at index (47, 190) (up to 0.001 allowed)
FAILED operators/test_matmul.py::test_op[128-128-32-1-4-2-107-233-83-True-True-float16-float16-True-True-None-None] - AssertionError: Tensor-likes are not close!

Mismatched elements: 384 / 24931 (1.5%)
Greatest absolute difference: 0.8779296875 at index (54, 183) (up to 1e-05 allowed)
Greatest relative difference: 7.625 at index (46, 57) (up to 0.001 allowed)
FAILED operators/test_matmul.py::test_op[128-128-32-1-4-2-107-233-83-True-True-bfloat16-bfloat16-True-True-None-None] - AssertionError: Tensor-likes are not close!

Mismatched elements: 384 / 24931 (1.5%)
Greatest absolute difference: 0.87890625 at index (54, 183) (up to 1e-05 allowed)
Greatest relative difference: 7.59375 at index (46, 57) (up to 0.016 allowed)
FAILED operators/test_matmul.py::test_op[128-256-32-1-8-2-None-None-None-True-False-float16-int8-True-True-None-None] - AssertionError: Tensor-likes are not close!

Mismatched elements: 384 / 32768 (1.2%)
Greatest absolute difference: 197.5 at index (85, 193) (up to 1e-05 allowed)
Greatest relative difference: 263.5 at index (70, 197) (up to 0.001 allowed)
FAILED operators/test_matmul.py::test_op[128-256-32-1-8-2-None-None-None-True-False-float16-int8-True-False-None-None] - AssertionError: Tensor-likes are not close!

Mismatched elements: 384 / 32768 (1.2%)
Greatest absolute difference: 197.5 at index (85, 193) (up to 1e-05 allowed)
Greatest relative difference: 263.5 at index (70, 197) (up to 0.001 allowed)
============ 6 failed, 699 passed, 236 skipped in 223.05s (0:03:43) ============
```

**Comment by chengjunlu** [comment]
The test case test_functional_regressions fails on DPAS.

```
test/regression/test_functional_regressions.py:10 (test_chained_matmul)
tensor([[1150., 1183.,  930.,  ..., 1197., 1038., 1068.],
        [1109., 1099.,  865.,  ..., 1104.,  972.,  972.],
        [1138., 1161.,  900.,  ..., 1157., 1010., 1008.],
        ...,
        [1267., 1278., 1016.,  ..., 1279., 1124., 1117.],
        [1104., 1109.,  886.,  ..., 1146.,  987., 1001.],
        [1056., 1065.,  848.,  ..., 1072.,  936.,  935.]], device='xpu:0',
       dtype=torch.float16) != tensor([[219., 283., 249.,  ..., 285., 138., 252.],
        [210., 278., 240.,  ..., 272., 135., 229.],
        [  0.,   0.,   0.,  ...,   0.,   0.,   0.],
        ...,
        [  0.,   0.,   0.,  ...,   0.,   0.,   0.],
        [  0.,   0.,   0.,  ...,   0.,   0.,   0.],
        [  0.,   0.,   0.,  ...,   0.,   0.,   0.]], device='xpu:0',
       dtype=torch.float16)
```

File: include/triton/Dialect/TritonIntelGPU/IR/TritonIntelGPUAttrDefs.td [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+}];
+
+  let parameters = (
+    ins
+    "unsigned":$repeatCount,
+    "unsigned":$systolicDepth,
+    "unsigned":$executionSize,
+    "unsigned":$opsPerChan,
+    ArrayRefParameter<"unsigned">:$warpsPerCTA__,
+    "unsigned":$sugGroupSize
```
Comment (chengjunlu) [inline include/triton/Dialect/TritonIntelGPU/IR/TritonIntelGPUAttrDefs.td:73]: It is duplicated with the attributed in the module.

File: lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandDPAS.cpp [hunk 2]
```diff
@@ -80,8 +81,7 @@ DpasMatmulLoader<opIdx>::computeLdsMatOffs(Value warpId, Value laneId,
   unsigned systolicDepth = dpasLayout.getSystolicDepth();
   unsigned repeatCount = dpasLayout.getRepeatCount();
   unsigned executionSize = dpasLayout.getExecutionSize();
-  unsigned opsPerChannel =
-      dpasLayout.getOpsPerChannel(tensorTy.getElementType());
+  unsigned opsPerChannel = dpasLayout.getOpsPerChan();
```
Comment (chengjunlu) [inline lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandDPAS.cpp:84]: Changes it to `getOpsPerChannel`

File: lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandDPAS.cpp [hunk 3]
```diff
@@ -279,16 +280,20 @@ Value loadOperand(ConversionPatternRewriter &rewriter, Location loc,
...
-  SmallVector<int64_t> numReps = encoding.getDPASRep(tensorShape, elemTy);
+  SmallVector<int64_t> elemsPerInstr;
+  if constexpr (opIdx == 0) {
+    auto shapeA = dpasLayout.getShapeA();
+    elemsPerInstr = {shapeA[0], shapeA[1]};
+  } else {
+    auto shapeB = dpasLayout.getShapeB();
+    elemsPerInstr = {shapeB[0], shapeB[1]};
+  }
+  SmallVector<int64_t> numReps = dpasLayout.getXMXRep(tensorShape, opIdx);
```
Comment (chengjunlu) [inline lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandDPAS.cpp:296]: Changes it to `getDPASRep`

File: lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/DPAS.cpp [hunk 4]
```diff
@@ -12,10 +13,25 @@
...
 using namespace mlir::triton;
 using mlir::triton::gpu::DotOperandEncodingAttr;
-using mlir::triton::gpu::DpasEncodingAttr;
+using mlir::triton::gpu::intel::DpasEncodingAttr;
 
 namespace {
 
+// data type for D_C_A_B.
+enum class DPASEngineType : uint8_t {
+  // floating-point XMX engine instr
```
Comment (chengjunlu) [inline lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/DPAS.cpp:22]: Change all the 'XMX' to 'DPAS'

File: lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/DPAS.cpp [hunk 5]
```diff
@@ -40,58 +183,73 @@ class DotOpDPASConversionHelper {
...
-        AEncoding.getDPASRep(ATensorTy.getShape(), ATensorTy.getElementType());
+        ADpasEncoding.getXMXRep(ATensorTy.getShape(), AEncoding.getOpIdx());
     auto repB =
-        BEncoding.getDPASRep(BTensorTy.getShape(), BTensorTy.getElementType());
+        BDpasEncoding.getXMXRep(BTensorTy.getShape(), BEncoding.getOpIdx());
     assert(repA[1] == repB[0] && "Unexpected rep for A and B operands");
 
     unsigned repM = repA[0], repN = repB[1], repK = repA[1];
 
+    auto mmaType = getMmaType(op);
```
Comment (chengjunlu) [inline lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/DPAS.cpp:199]: Use the static type instead of the auto.

**Review by pengtu — COMMENTED** [review]
Use the get_current_target()[1] to set the Intel GPU architecture before IPEX pass the correct capability to Triton

**Comment by chengjunlu** [comment]
This is obsoleted

**Pull request closed without merging**